src/Libraries/Microsoft.Extensions.AI/ChatCompletion/AnonymousChatClient.cs

```diff
@@ -0,0 +1,242 @@
+        }
+        else if (_completeStreamingFunc is not null)
+        {
+            return _completeStreamingFunc(arguments);
+        }
+        else
+        {
+            Debug.Assert(_completeFunc is not null, "Expected non-null non-streaming delegate.");
+            return CompleteStreamingAsyncViaCompleteAsync(_completeFunc!(arguments));
+
+            static async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsyncViaCompleteAsync(Task<ChatCompletion> task)
+            {
+                ChatCompletion completion = await task.ConfigureAwait(false);
+                foreach (var update in completion.ToStreamingChatCompletionUpdates())
+                {
+                    yield return update;
+                }
+            }
+        }
+    }
+
+    /// <summary>Throws an exception if both of the specified delegates are null.</summary>
+    /// <exception cref="ArgumentNullException">Both <paramref name="completeFunc"/> and <paramref name="completeStreamingFunc"/> are <see langword="null"/>.</exception>
+    internal static void ThrowIfBothDelegatesNull(
+        Func<Arguments, Task<ChatCompletion>>? completeFunc,
+        Func<Arguments, IAsyncEnumerable<StreamingChatCompletionUpdate>>? completeStreamingFunc)
+    {
+        if (completeFunc is null && completeStreamingFunc is null)
+        {
+            Throw.ArgumentNullException(nameof(completeFunc), $"At least one of the {nameof(completeFunc)} or {nameof(completeStreamingFunc)} delegates must be non-null.");
+        }
+    }
+
+    /// <summary>Represents the arguments passed to the delegates used by <see cref="AnonymousChatClient"/>.</summary>
+    public sealed class Arguments
+    {
+        /// <summary>Initializes a new instance of the <see cref="Arguments"/> class.</summary>
+        internal Arguments(
+            IList<ChatMessage> chatMessages,
+            ChatOptions? options,
+            CancellationToken cancellationToken,
+            IChatClient innerClient,
+            IServiceProvider? services)
+        {
+            ChatMessages = chatMessages;
+            Options = options;
+            CancellationToken = cancellationToken;
+            InnerClient = innerClient;
+            Services = services;
+        }
+
+        /// <summary>Gets the chat content to send.</summary>
+        public IList<ChatMessage> ChatMessages { get; }
+
+        /// <summary>Gets the chat options to configure the request.</summary>
+        public ChatOptions? Options { get; }
+
+        /// <summary>Gets the <see cref="CancellationToken"/> to monitor for cancellation requests.</summary>
+        public CancellationToken CancellationToken { get; }
+
+        /// <summary>Gets the inner client to which to delegate.</summary>
+        public IChatClient InnerClient { get; }
+
+        /// <summary>Gets any services associated with the chat client.</summary>
+        public IServiceProvider? Services { get; }
```

Comment:
I'm guessing by the presence of this parameter that the inner service provider can differ from the outer one?

Reply:
I'm not sure what you mean. The services here are passed in from the builder.

Reply:
You can also access the services on the `InnerClient` property.

Reply:
Are you referring to GetService? That's not the same. It does not provide access to IServiceProvider.